pwiz_tools/Skyline/TestRunnerLib/RunTests.cs

```diff
@@ -517,34 +530,98 @@ public bool Run(TestInfo test, int pass, int testNumber, string dmpDir, bool hea
+                //      c:\crazy long username\massive subdir name\wacky installation dirnamne\pwiz_tools\Skyline\~t&mp ^\TMMENF910 试验"
+                // instead of
+                //      c:\crazy long username\massive subdir name\wacky installation dirnamne\pwiz_tools\Skyline\SkylineTester temp&di^rs\TestMyMostExcellentNebulousFunction 试验"
+                tmpTestDir = Path.GetFullPath(Path.Combine(testDir, @"..", @"~t&mp ^",
+                    $"{string.Concat(test.TestMethod.Name.Where(char.IsUpper))}{test.TestMethod.Name.Sum(c => c)}{unicode}"));
+            }
+
+            if (!Directory.Exists(tmpTestDir))
+            {
+                Directory.CreateDirectory(tmpTestDir);
+            }
+
+            Environment.SetEnvironmentVariable(@"TMP", tmpTestDir);
+
+            // Decorate tempfile names with peculiar characters
+            PathEx.RandomFileNameDecoration = @$"t^m&p{unicode} ";
+            return tmpTestDir;
+        }
+
+        public static void CleanupMSAmandaTmpFiles()
+        {
+            // MSAmanda intentionally leaves tempfiles behind (as caches in case of repeat runs)
+            // But our test system wants a clean finish
+            // TODO(MattC): tidy up MSAmanda implementation so that we can distinguish intentional uses of tmp dir (caching potentially re-used files) from accidental directory creation and/or not-reused files within
+            var msAmandaTmpDir = Path.Combine(Path.GetTempPath(), @"~SK_MSAmanda" /* must match MSAmandaSearchWrapper.MS_AMANDA_TMP */);
+            try
+            {
+                Directory.Delete(msAmandaTmpDir, true);
+            }
+            catch
+            {
+                // ignored
+            }
+        }
+
         /// <summary>
         /// Resets the test directory to empty when cleanupLevel is 'all'.
         /// </summary>
         /// <returns>A list of the entries, if any, that were removed</returns>
-        private List<string> CleanUpTestDir()
+        private List<string> CleanUpTestDir(string tmpTestDir, bool final)
         {
-            var allEntries = new List<string>();
+            CleanupMSAmandaTmpFiles();  // TODO(MattC): tidy up MSAmanda implementation so that we can distinguish intentional uses of tmp dir (caching potentially re-used files) from accidental directory creation and/or not-reused files within
+            var abandonedFilesList = new List<string>();
             // If everything is supposed to be cleaned up, then check for any left over files
-            if (_cleanupLevelAll && Directory.Exists(TestContext.TestDir))
+            if (_cleanupLevelAll)
             {
-                allEntries.AddRange(Directory.EnumerateFileSystemEntries(TestContext.TestDir));
-                if (allEntries.Count > 0)
+                CleanupAbandonedFiles(TestContext.TestDir, true, abandonedFilesList);
+                CleanupAbandonedFiles(tmpTestDir, !final, abandonedFilesList);
```

Comment:
I guess I wonder about this placement. It would seem to imply that we are fine with leaving temp files around when _cleanupLevelAll is false, as we really are with leaving around the files tests produce, since they can be very helpful artifacts when a test fails. I could imagine we would want to see the temp files go away no matter what the clean-up level. Are we expecting normal use to leave temp files around?

Reply:
It would not be atypical program behavior at least. Just look at all the cruft that builds up in %TEMP%. :) But I think if Skyline is creating a temp file that is supposed to be cleaned up during execution but maybe isn't due to a bug or error, it's ok to clean it up always. If we want to keep a file around for understanding why an error is happening, it's better to have it be a non-temp file (like I will move the DDA params files to the output directory instead of TEMP).

Reply:
Right. I think that means that the tmpTestDir check should go outside the if block, so that we check for clean-up regardless of whether the test settings are set to clean up all files.

Reply:
I agree.